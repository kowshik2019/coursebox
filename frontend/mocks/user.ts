export const mockUser = {
  jwt: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NSwiaWF0IjoxNjUwOTAxNjEwLCJleHAiOjE2NTM0OTM2MTB9.5XCgAR25c3AH_09Jv31CWNF-48mpRqGUat79XGYuNJo",
  user: {
    id: 5,
    username: "John Doe",
    email: "john@doe.com",
    password: "testpass123!",
    provider: "local",
    confirmed: true,
    blocked: false,
    createdAt: "2022-04-20T12:58:41.661Z",
    updatedAt: "2022-04-20T12:58:41.661Z",
  },
};
