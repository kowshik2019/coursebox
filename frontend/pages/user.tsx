import type { NextPage } from "next";
import { useRouter } from "next/router";
import useSWR from "swr";
import { useSelector, useDispatch } from "react-redux";

import { RootState, AppDispatch } from "@/store";
import { logout, UserState, selectUser } from "@/services/userSlice";

import { Button } from "@/components/Button";
import { CenteredTile } from "@/components/Tile";

const User: NextPage = () => {
  const router = useRouter();
  const { data, error } = useSWR("/users/me");

  const dispatch = useDispatch<AppDispatch>();
  const user = useSelector<RootState, UserState>(selectUser);

  const logoutHandler = async () => {
    await dispatch(logout());
    router.push("/");
  };

  if (user.error || data?.error || error) {
    dispatch(logout());
    router.push("/login");
  }

  return data ? (
    <CenteredTile header="Profile">
      <h3>username: {data?.username}</h3>
      <h3>email: {data?.email}</h3>
      <Button onClick={logoutHandler}>Logout</Button>
    </CenteredTile>
  ) : null;
};

export default User;
