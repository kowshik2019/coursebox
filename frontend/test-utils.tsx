import { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { Themes } from "./styles/themes";
import { store } from "./store";

const Wrapper: FC = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
  </Provider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
