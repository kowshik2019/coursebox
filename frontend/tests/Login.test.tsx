import { useRouter } from "next/router";

import { pageRender as render, screen, act, waitFor } from "@/test-utils";
import userEvent from "@testing-library/user-event";

import Login from "@/pages/login";

import { mockUser } from "@/mocks/user";

jest.mock("next/router", () => ({
  ...jest.requireActual("next/router"),
  useRouter: jest.fn(),
}));

describe("Login page", () => {
  it("Render check", () => {
    const { container } = render(<Login />);
    expect(container).toMatchSnapshot();
  });
  it("Client validation check", async () => {
    render(<Login />);

    const submitButton = screen.getByRole("button", { name: "Sign In" });

    userEvent.type(screen.getByRole("textbox", { name: "Identifier" }), "test");
    userEvent.type(screen.getByLabelText("password"), "test");

    await act(async () => {
      userEvent.click(submitButton);
    });

    expect(screen.getByText("Min length 6!")).toBeInTheDocument();
    expect(screen.getByText("Min length 8!")).toBeInTheDocument();
  });
  it("Server validation check", async () => {
    render(<Login />);

    const submitButton = screen.getByRole("button", { name: "Sign In" });

    userEvent.type(
      screen.getByRole("textbox", { name: "Identifier" }),
      "test@test.test"
    );
    userEvent.type(screen.getByLabelText("password"), "testpassworddd!");

    await act(async () => {
      userEvent.click(submitButton);
    });

    expect(
      await screen.findByText("Invalid identifier or password")
    ).toBeInTheDocument();
  });
  it("Successful login check", async () => {
    // Mock the router
    const push = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push });

    render(<Login />);

    const submitButton = screen.getByRole("button", { name: "Sign In" });

    userEvent.type(
      screen.getByRole("textbox", { name: "Identifier" }),
      mockUser.user.email
    );

    userEvent.type(screen.getByLabelText("password"), mockUser.user.password);

    await act(async () => {
      userEvent.click(submitButton);
    });

    // Check if the user is redirected to the user page
    await waitFor(() => expect(push).toHaveBeenCalledWith("/user"));
  });
});
