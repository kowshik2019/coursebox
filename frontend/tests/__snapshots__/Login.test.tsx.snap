// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page Render check 1`] = `
<div>
  <form>
    <div
      class="css-1oczm9i"
    >
      <section
        class="css-eatc9a"
      >
        <h2>
          Login
        </h2>
        <h3>
          <span
            class="css-sk8pyf"
          >
             Â 
          </span>
          <span
            class="css-sk8pyf"
          />
          <span
            class="css-sk8pyf"
            data-testid="hidden-request-state"
            style="visibility: hidden;"
          />
        </h3>
        <label
          class="css-5h33el"
          height="7"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            Identifier
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              minlength="6"
              name="identifier"
              placeholder="username or email"
            />
          </div>
          <span
            class="css-1reg724"
          />
        </label>
        <label
          class="css-5h33el"
          height="7"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            password
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              minlength="6"
              name="password"
              placeholder="password"
              type="password"
            />
          </div>
          <span
            class="css-1reg724"
          />
        </label>
        <button
          class="css-1rttvxo"
          color="primary"
          type="submit"
        >
          Sign In
        </button>
        <h3>
          <a
            class="css-pm0vza"
            href="/registration"
          >
            Create account
          </a>
        </h3>
      </section>
    </div>
  </form>
</div>
`;
